### ISSUE TITLE
Equal and IEqualityComparer

### ISSUE THREAD

**Issue by NikitaEgorov** [FluentValidation/FluentValidation#1642]
**Is your feature request related to a problem? Please describe.**
When use Equal with custom comparer I should implement non-generic `IEqualityComparer`

**Describe the solution you'd like**
Use `IEqualityComparer<T>`

```csharp
// My rule
this.RuleFor(x => x.StartDate).Equal(x => x.EndDate, new MonthComparer());

// My comparer
private class MonthComparer : IEqualityComparer
{
    public new bool Equals(object x, object y)
    {
        var xx = (DateTime)x;
        var yy = (DateTime)y;
        return xx.Month == yy.Month && xx.Year == yy.Year;
    }

    public int GetHashCode(object obj)
    {
        var date = (DateTime)obj;
        return HashCode.Combine(date.Month, date.Year);
    }
}


// Comparer what I want
private class MonthComparer : IEqualityComparer<DateTime>, IEqualityComparer
{
    public bool Equals(DateTime x, DateTime y) => x.Month == y.Month && x.Year == y.Year;

    public int GetHashCode(DateTime obj) => HashCode.Combine(obj.Month, obj.Year);
}

// but, generic IEqualityComparer doesn't have base interface 
namespace System.Collections.Generic
{
  public interface IEqualityComparer<in T>
  {
    bool Equals([AllowNull] T x, [AllowNull] T y);

    int GetHashCode([DisallowNull] T obj);
  }
}





```

**Comment by JeremySkinner** [comment]
This will be done as part of our 10.0 release (in 9.x and older FluentValidation’s internal model is non-generic)

**Comment by JeremySkinner** [comment]
Implemented for 10.0
